File: pkg/geometry/plane.go
package geometry

import (
	"math"

	"sif/gogs/eric/relray/pkg/vec"
)

// Plane is an infinite plane on the XZ plane (Y=0) with normal pointing +Y.
// Use a Transformed wrapper to position and orient it in the scene.
type Plane struct{}

func (p *Plane) Intersect(origin, dir vec.Vec3, tMin, tMax float64) (Hit, bool) {
	if math.Abs(dir.Y) < 1e-12 {
		return Hit{}, false // ray parallel to plane
	}
	t := -origin.Y / dir.Y
	if t < tMin || t > tMax {
		return Hit{}, false
	}
	pt := origin.Add(dir.Scale(t))
	h := Hit{T: t, Point: pt}
	h.SetFaceNormal(dir, vec.Vec3{Y: 1})
	return h, true
}

File: rust/src/geometry/plane.rs
use crate::vec::Vec3;
use super::{Hit, Shape};

pub struct Plane;

impl Shape for Plane {
    fn intersect(&self, origin: Vec3, dir: Vec3, t_min: f64, t_max: f64) -> Option<Hit> {
        if dir.y.abs() < 1e-12 {
            return None;
        }
        let t = -origin.y / dir.y;
        if t < t_min || t > t_max {
            return None;
        }
        let pt = origin + dir * t;
        let mut h = Hit { t, point: pt, ..Default::default() };
        h.set_face_normal(dir, Vec3 { x: 0.0, y: 1.0, z: 0.0 });
        Some(h)
    }
}
